sql/src/main/java/io/druid/sql/calcite/aggregation/ApproxCountDistinctSqlAggregator.java

```diff
@@ -0,0 +1,122 @@
+import io.druid.query.aggregation.hyperloglog.HyperUniqueFinalizingPostAggregator;
+import io.druid.query.aggregation.hyperloglog.HyperUniquesAggregatorFactory;
+import io.druid.query.dimension.DimensionSpec;
+import io.druid.segment.column.ValueType;
+import io.druid.sql.calcite.expression.Expressions;
+import io.druid.sql.calcite.expression.RowExtraction;
+import io.druid.sql.calcite.table.RowSignature;
+import org.apache.calcite.rel.core.AggregateCall;
+import org.apache.calcite.rel.core.Project;
+import org.apache.calcite.sql.SqlAggFunction;
+import org.apache.calcite.sql.SqlFunctionCategory;
+import org.apache.calcite.sql.SqlKind;
+import org.apache.calcite.sql.type.InferTypes;
+import org.apache.calcite.sql.type.OperandTypes;
+import org.apache.calcite.sql.type.ReturnTypes;
+import org.apache.calcite.sql.type.SqlTypeName;
+
+import java.util.List;
+
+public class ApproxCountDistinctSqlAggregator implements SqlAggregator
+{
+  private static final SqlAggFunction FUNCTION_INSTANCE = new ApproxCountDistinctSqlAggFunction();
+  private static final String NAME = "APPROX_COUNT_DISTINCT";
+
+  @Override
+  public SqlAggFunction calciteFunction()
+  {
+    return FUNCTION_INSTANCE;
+  }
+
+  @Override
+  public Aggregation toDruidAggregation(
+      final String name,
+      final RowSignature rowSignature,
+      final List<Aggregation> existingAggregations,
+      final Project project,
+      final AggregateCall aggregateCall
+  )
+  {
+    final RowExtraction rex = Expressions.toRowExtraction(
+        rowSignature.getRowOrder(),
+        Expressions.fromFieldAccess(
+            rowSignature,
+            project,
+            Iterables.getOnlyElement(aggregateCall.getArgList())
+        )
+    );
+    if (rex == null) {
+      return null;
+    }
+
+    final AggregatorFactory aggregatorFactory;
+
+    if (rowSignature.getColumnType(rex.getColumn()) == ValueType.COMPLEX) {
+      aggregatorFactory = new HyperUniquesAggregatorFactory(name, rex.getColumn());
+    } else {
+      final DimensionSpec dimensionSpec = rex.toDimensionSpec(rowSignature, null);
+      if (dimensionSpec == null) {
+        return null;
+      }
+
+      aggregatorFactory = new CardinalityAggregatorFactory(name, ImmutableList.of(dimensionSpec), false);
+    }
+
+    return Aggregation.createFinalizable(
+        ImmutableList.<AggregatorFactory>of(aggregatorFactory),
+        null,
+        new PostAggregatorFactory()
+        {
+          @Override
+          public PostAggregator factorize(String outputName)
+          {
+            return new HyperUniqueFinalizingPostAggregator(outputName, name);
+          }
+        }
+    );
+  }
+
+  private static class ApproxCountDistinctSqlAggFunction extends SqlAggFunction
+  {
+    ApproxCountDistinctSqlAggFunction()
+    {
+      super(
+          NAME,
+          null,
+          SqlKind.OTHER_FUNCTION,
+          ReturnTypes.explicit(SqlTypeName.BIGINT),
+          InferTypes.VARCHAR_1024,
```

Comment:
No, it's just there for the benefit of Calcite's SQL validator. It's not going to affect query processing.